packages/jupyter-ai/jupyter_ai/personas/directories.py

```diff
@@ -0,0 +1,98 @@
+from pathlib import Path
+from typing import Optional
+
+
+def find_dot_dir(
+    dir: str, dot_dir: str, root_dir: Optional[str] = None
+) -> Optional[str]:
+    """
+    Find the nearest dot directory by traversing up from the given directory.
+
+    Args:
+        dir (str): The starting directory path
+        dot_dir (str): The dot directory name to search for (e.g., '.jupyter', '.git')
+        root_dir (Optional[str]): The root directory to stop searching at.
+            If None, searches to filesystem root.
+
+    Returns:
+        str: The absolute path to the dot directory if found, None otherwise
+
+    Raises:
+        ValueError: If dir is not a directory
+    """
+    current_path = Path(dir).resolve()
+
+    # Validate that the starting path is a directory
+    if not current_path.is_dir():
+        raise ValueError(f"'{dir}' is not a directory")
+
+    # Determine the stopping point
+    if root_dir is not None:
+        stop_path = Path(root_dir).resolve()
+    else:
+        stop_path = current_path.parent  # Will be set to filesystem root in the loop
+
+    while current_path != current_path.parent:  # Stop at root directory
+        target_dir = current_path / dot_dir
+        try:
+            if target_dir.is_dir():
+                return str(target_dir)
+        except PermissionError:
+            # Stop searching if we don't have permission to access this directory
+            break
+
+        # Stop if we've reached the specified root directory
+        if root_dir is not None and current_path == stop_path:
+            break
+
+        current_path = current_path.parent
+
+    return None
+
+
+def find_workspace_dir(dir: str, root_dir: Optional[str] = None) -> str:
+    """
+    Find the workspace directory using the following algorithm:
+
+    1. First finds the nearest .jupyter directory
+    2. If the .jupyter directory has a parent that is the root_dir,
+       go to step 3, otherwise return the .jupyter directory
+    3. Try to find a .git directory, if found return that
+    4. Simply return the directory the chat file is in.
```

Comment:
1. What is the intended definition & purpose of a "workspace"? 

2. Intuitively, I expected that workspaces would be defined as the contents root, since that is the "main directory" you see when starting JupyterLab. Why does this return `.jupyter` or `.git`, while doing so on the condition that `.jupyter`'s parent cannot be the contents root?

Reply:
The workspace is intended to function similar to show it does for Desktop IDEs. The problem is that unlike desktop IDEs, many people run JupyterLab against their entire home directory. We want to avoid (if at all possible) treating the entire file system as the workspace as code reading/understanding agents will end up walking the entire file system. The logic here tries to establish a heuristic that makes a reasonable set of decisions based on a combination of .jupyter directories, .git directories, and the CWD of the chat file. We can modify and tune this logic over time within this API, but we need something to get started with.